pennylane/devices/qubit/simulate.py

```diff
@@ -335,6 +360,361 @@ def simulate_partial(k):
+    sequence corresponds to `[0,1,1,0,0,1]`.
+    """
+    if mcm_active:
+        shape = next(iter(mcm_samples.values())).shape
+        mask = np.ones(shape, dtype=bool)
+        for k, v in mcm_active.items():
+            if k == op:
+                break
+            mask = np.logical_and(mask, mcm_samples[k] == v)
+        if op not in mcm_samples:
+            mcm_samples[op] = np.empty(shape, dtype=samples.dtype)
+        mcm_samples[op][mask] = samples
+    else:
+        mcm_samples[op] = samples
+
+
+def circuit_up_to_first_mcm(circuit):
+    """Returns two circuits; one that runs up-to the next mid-circuit measurement and one that runs beyond it.
+
+    Measurement processes are computed on each branch, and then combined at the node.
+    This can be done recursively until a single node is left.
+    This is true for `counts`, `expval`, `probs` and `sample` but not `var` measurements.
+    There is no way to recombine "partial variances" from two branches, so `var` measurements are replaced
+    by `sample` measurements from which the variance is calculated (once samples from all branches are available).
+
+    Args:
+        circuit (QuantumTape): The circuit to simulate
+
+    Returns:
+        QuantumTape: Circuit up to the first MCM and measuring the MCM samples if an MCM is found and ``circuit`` otherwise
+        (QuantumTape, None): Rest of the circuit
+        (MidMeasureMP, None): The first MCM encountered in the circuit
+    """
+
+    # find next MidMeasureMP
+    def find_next_mcm(circuit):
+        for i, op in enumerate(circuit.operations):
+            if isinstance(op, MidMeasureMP):
+                return i, op
+        return len(circuit.operations) + 1, None
+
+    i, op = find_next_mcm(circuit)
+
+    if op is None:
+        return circuit, None, None
+
+    # run circuit until next MidMeasureMP and sample
+    circuit_base = qml.tape.QuantumScript(
+        circuit.operations[0:i],
+        [qml.sample(wires=op.wires)],
+        shots=circuit.shots,
+        trainable_params=circuit.trainable_params,
+    )
+    # circuit beyond next MidMeasureMP with VarianceMP <==> SampleMP
+    new_measurements = []
+    for m in circuit.measurements:
+        if not m.mv:
+            if isinstance(m, VarianceMP):
+                new_measurements.append(SampleMP(obs=m.obs))
+            else:
+                new_measurements.append(m)
+    circuit_next = qml.tape.QuantumScript(
+        circuit.operations[i + 1 :],
+        new_measurements,
+        shots=circuit.shots,
+        trainable_params=circuit.trainable_params,
+    )
+    return circuit_base, circuit_next, op
+
+
+def measurement_with_no_shots(measurement):
+    """Returns a NaN scalar or array of the correct size when executing an all-invalid-shot circuit."""
+    if isinstance(measurement, ProbabilityMP):
+        if measurement.obs:
+            return np.nan * np.ones(measurement.obs.eigvals().size)
```

Comment:
When would this differ from 2**len(measurement.wires)?

Reply:
I think I thought `measurement.wires` would be defined only if we requested `qml.probs(wires=wires)`. But you're right, the attribute is properly set.